File: graphite-backend/app/routes/backup.py
# app/routes/backup.py
# 异步版本 - 支持任务状态轮询
# ✅ 符合项目规范：Blueprint定义时不设置url_prefix

from flask import Blueprint, request, jsonify, send_file
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.utils.decorators import admin_required
from app.utils.backup_manager import BackupManager

# ✅ 符合项目规范：不在这里设置url_prefix
backup_bp = Blueprint('backup', __name__)


@backup_bp.route('', methods=['POST'])
@jwt_required()
@admin_required()
def create_backup():
    """
    创建数据库备份（异步）
    
    立即返回任务ID，前端通过轮询查询任务状态
    
    Returns:
        201: {success: true, task_id: str, filename: str, message: str}
        500: {success: false, message: str}
    """
    try:
        current_user_id = int(get_jwt_identity())
        
        result = BackupManager.create_backup_async(current_user_id)
        
        if result['success']:
            return jsonify(result), 201
        else:
            return jsonify(result), 500
            
    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'备份失败: {str(e)}'
        }), 500


@backup_bp.route('/task/<task_id>', methods=['GET'])
@jwt_required()
@admin_required()
def get_task_status(task_id):
    """
    获取备份任务状态（用于前端轮询）
    
    Args:
        task_id: 任务ID
        
    Returns:
        200: {task_id, status, filename, file_size, error_message, ...}
        404: {error: '任务不存在'}
    """
    try:
        result = BackupManager.get_task_status(task_id)
        
        if 'error' in result:
            return jsonify(result), 404
        else:
            return jsonify(result), 200
            
    except Exception as e:
        return jsonify({
            'error': f'查询任务状态失败: {str(e)}'
        }), 500


@backup_bp.route('', methods=['GET'])
@jwt_required()
@admin_required()
def list_backups():
    """
    获取备份列表（包含任务状态）
    
    Returns:
        200: {backups: [], total: int}
        500: {error: str}
    """
    try:
        backups = BackupManager.list_backups()
        return jsonify({
            'backups': backups,
            'total': len(backups)
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': f'获取备份列表失败: {str(e)}'
        }), 500


@backup_bp.route('/<filename>', methods=['GET'])
@jwt_required()
@admin_required()
def download_backup(filename):
    """
    下载备份文件
    
    Args:
        filename: 备份文件名
        
    Returns:
        200: 文件流
        400: 非法的文件路径
        404: 文件不存在
        500: 下载失败
    """
    try:
        # ✅ 安全验证：只允许 .sql 文件
        if not filename.endswith('.sql'):
            return jsonify({'error': '只支持下载 .sql 文件'}), 400
        
        filepath = BackupManager.BACKUP_DIR / filename
        
        # 验证文件存在
        if not filepath.exists():
            return jsonify({'error': '备份文件不存在'}), 404
        
        # ✅ 安全检查：防止目录遍历攻击
        if not str(filepath.resolve()).startswith(str(BackupManager.BACKUP_DIR.resolve())):
            return jsonify({'error': '非法的文件路径'}), 400
        
        # 发送文件
        return send_file(
            filepath,
            as_attachment=True,
            download_name=filename,
            mimetype='application/sql'
        )
        
    except Exception as e:
        return jsonify({'error': f'下载失败: {str(e)}'}), 500


@backup_bp.route('/<filename>', methods=['DELETE'])
@jwt_required()
@admin_required()
def delete_backup(filename):
    """
    删除备份文件
    
    Args:
        filename: 备份文件名
        
    Returns:
        200: {success: true, message: str}
        400: {success: false, message: str}
        500: {error: str}
    """
    try:
        current_user_id = int(get_jwt_identity())
        
        result = BackupManager.delete_backup(filename, current_user_id)
        
        if result['success']:
            return jsonify(result), 200
        else:
            return jsonify(result), 400
            
    except Exception as e:
        return jsonify({
            'success': False,
            'error': f'删除失败: {str(e)}'
        }), 500


@backup_bp.route('/statistics', methods=['GET'])
@jwt_required()
@admin_required()
def get_statistics():
    """
    获取备份统计信息
    
    Returns:
        200: {total_backups, total_size, last_backup_time, database_size, running_tasks}
        500: {error: str}
    """
    try:
        stats = BackupManager.get_statistics()
        return jsonify(stats), 200
        
    except Exception as e:
        return jsonify({
            'error': f'获取统计信息失败: {str(e)}'
        }), 500

File: graphite-frontend/src/api/backup.ts
// src/api/backup.ts
// 数据库备份管理 API

import request from '@/utils/request'

// ========== 类型定义 ==========

/**
 * 备份任务状态
 */
export type BackupTaskStatus = 'pending' | 'running' | 'success' | 'failed'

/**
 * 创建备份响应
 */
export interface CreateBackupResponse {
  success: boolean
  task_id: string
  filename: string
  message: string
}

/**
 * 任务状态响应
 */
export interface TaskStatusResponse {
  task_id: string
  status: BackupTaskStatus
  filename: string
  file_size: number | null
  error_message: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
}

/**
 * 备份文件信息
 */
export interface BackupFile {
  filename: string
  file_size: number
  created_at: string
  status: BackupTaskStatus
  file_path: string
}

/**
 * 备份列表响应
 */
export interface BackupListResponse {
  backups: BackupFile[]
  total: number
}

/**
 * 备份统计信息
 */
export interface BackupStatistics {
  total_backups: number      // 备份文件数量
  total_size: number          // 总占用空间（字节）
  last_backup_time: string | null  // 最后备份时间
  database_size: number       // 数据库大小（字节）
  running_tasks: number       // 正在运行的任务数
}

// ========== API 函数 ==========

/**
 * 创建数据库备份（异步）
 *
 * @returns 任务ID和文件名
 */
export function createBackup() {
  return request<CreateBackupResponse>({
    url: '/admin/backup',
    method: 'post'
  })
}

/**
 * 查询备份任务状态
 *
 * @param taskId 任务ID
 * @returns 任务状态详情
 */
export function getTaskStatus(taskId: string) {
  return request<TaskStatusResponse>({
    url: `/admin/backup/task/${taskId}`,
    method: 'get'
  })
}

/**
 * 获取备份列表
 *
 * @returns 备份文件列表
 */
export function getBackupList() {
  return request<BackupListResponse>({
    url: '/admin/backup',
    method: 'get'
  })
}

/**
 * 下载备份文件
 *
 * @param filename 文件名
 * @returns Blob下载流
 */
export function downloadBackup(filename: string) {
  return request({
    url: `/admin/backup/${filename}`,
    method: 'get',
    responseType: 'blob'
  })
}

/**
 * 删除备份文件
 *
 * @param filename 文件名
 * @returns 删除结果
 */
export function deleteBackup(filename: string) {
  return request<{ success: boolean; message: string }>({
    url: `/admin/backup/${filename}`,
    method: 'delete'
  })
}

/**
 * 获取备份统计信息
 *
 * @returns 统计数据
 */
export function getBackupStatistics() {
  return request<BackupStatistics>({
    url: '/admin/backup/statistics',
    method: 'get'
  })
}

// ========== 工具函数 ==========

/**
 * 轮询任务状态直到完成
 *
 * @param taskId 任务ID
 * @param onUpdate 状态更新回调
 * @param interval 轮询间隔（毫秒），默认3000
 * @returns Promise，在任务完成或失败时resolve
 */
export function pollTaskStatus(
  taskId: string,
  onUpdate: (status: TaskStatusResponse) => void,
  interval: number = 3000
): Promise<TaskStatusResponse> {
  return new Promise((resolve, reject) => {
    const poll = async () => {
      try {
        const status = await getTaskStatus(taskId)

        // 调用更新回调
        onUpdate(status)

        // 检查任务是否完成
        if (status.status === 'success' || status.status === 'failed') {
          resolve(status)
        } else {
          // 继续轮询
          setTimeout(poll, interval)
        }
      } catch (error) {
        reject(error)
      }
    }

    // 开始轮询
    poll()
  })
}

/**
 * 格式化文件大小
 *
 * @param bytes 字节数
 * @returns 格式化的大小字符串（如 "1.23 MB"）
 */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 B'

  const k = 1024
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB']
  const i = Math.floor(Math.log(bytes) / Math.log(k))

  return `${(bytes / Math.pow(k, i)).toFixed(2)} ${sizes[i]}`
}

/**
 * 格式化日期时间
 *
 * @param dateString ISO日期字符串
 * @returns 格式化的日期时间（如 "2024-12-16 21:52:43"）
 */
export function formatDateTime(dateString: string | null): string {
  if (!dateString) return '-'

  const date = new Date(dateString)
  const year = date.getFullYear()
  const month = String(date.getMonth() + 1).padStart(2, '0')
  const day = String(date.getDate()).padStart(2, '0')
  const hours = String(date.getHours()).padStart(2, '0')
  const minutes = String(date.getMinutes()).padStart(2, '0')
  const seconds = String(date.getSeconds()).padStart(2, '0')

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`
}

/**
 * 触发浏览器下载
 *
 * @param blob Blob对象
 * @param filename 文件名
 */
export function triggerDownload(blob: Blob, filename: string) {
  const url = window.URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
  window.URL.revokeObjectURL(url)
}
